Optimize cleanup compaction get ranges for invalidation

bhalevy opened the pull request:
Take advantage of the facts that both the owned ranges
and the initial non_owned_ranges (derived from the set of sstables)
are deoverlapped and sorted by start token to turn
the calculation of the final non_owned_ranges from
quadratic to linear.

Fixes #11922

bhalevy commented:
Converted to draft as this PR depends on https://github.com/scylladb/scylladb/pull/11902

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/2860/

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/2876/

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/2886/

raphaelsc commented on compaction/compaction.cc:
```diff
@@ -1212,18 +1212,57 @@ class cleanup_compaction final : public regular_compaction {
         // optimize set of potentially overlapping ranges by deoverlapping them.
         non_owned_ranges = dht::partition_range::deoverlap(std::move(non_owned_ranges), dht::ring_position_comparator(*_schema));
 
```
I think we need to document dht::partition_range::deoverlap() (above the function declaration) returns a sorted vector, such that no change in the future breaks this assumption.

bhalevy commented:
In [7793003](https://github.com/scylladb/scylladb/commit/779300320fb1f2b062279e5659f3eda53839d8cd):
```diff
diff --git a/interval.hh b/interval.hh
index cf33b1e0c..b6837f6b6 100644
--- a/interval.hh
+++ b/interval.hh
@@ -585,7 +585,8 @@ class nonwrapping_interval {
         return _interval == other._interval;
     }
     // Takes a vector of possibly overlapping intervals and returns a vector containing
-    // a set of non-overlapping intervals covering the same values.
+    // a set of non-overlapping intervals covering the same values,
+    // sorted by the interval start bound.
     template<IntervalComparatorFor<T> Comparator, typename IntervalVec>
     requires requires (IntervalVec vec) {
         { vec.begin() } -> std::random_access_iterator;
```

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/2891/

denesb commented on compaction/compaction_manager.cc:
```diff
@@ -1470,8 +1470,8 @@ class compaction_manager::cleanup_sstables_compaction_task : public compaction_m
+    const auto& first = sst->get_first_partition_key();
+    const auto& last = sst->get_last_partition_key();
     auto first_token = dht::get_token(*s, first);
     auto last_token = dht::get_token(*s, last);
```
You can further optimize this by using `sst->get_first_decorated_key()` and use not re-decorate the key here.

bhalevy replied:
That's possible if the partitioner didn't change (which is most likely).
But I'll add a check for that, just in case.

bhalevy commented:
In v2 ([559a688](https://github.com/scylladb/scylladb/commit/559a6880a648f71f561caeb4aeb0367f9d02a208)):
- compaction_manager: needs_cleanup: get first/last tokens from sstable decorated keys
- rebased + retested

denesb commented on compaction/compaction_manager.cc:
```diff
@@ -1470,10 +1470,18 @@ class compaction_manager::cleanup_sstables_compaction_task : public compaction_m
+    dht::token first_token;
+    dht::token last_token;
+    // avoid partitioner string compare if schema is the same
+    if (sst->get_schema() == s || sst->get_schema()->get_partitioner() == s->get_partitioner()) [[likely]] {
```
I don't think we even support changing the partitioner currently. It is hardcoded into the schema.

bhalevy replied:
so maybe we should just a debug level assert to cement this assumption

denesb replied:
We rely on this fact throughout the codebase. I just don't think we should start to add asserts for it now. It is like adding an assert that the primary key type didn't change. It is pointless IMO.
Also, these keys are decorated after the sstables are loaded, they are not stored in a decorated form on disk.

bhalevy replied:
okay, I'll simplify this patch then.

bhalevy replied:
Done in v4.

bhalevy commented:
In v4 ([f5e4427](https://github.com/scylladb/scylladb/commit/f5e44272c333726cd8918d71b62a8b6bf50f7005)):
- simplified "compaction_manager: needs_cleanup: get first/last tokens from sstable decorated keys"

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/2915/

raphaelsc commented on compaction/compaction.cc:
```diff
@@ -1212,18 +1212,57 @@ class cleanup_compaction final : public regular_compaction {
+            case 0:
+                // current non_owned_range is fully owned, done with it
+                ++non_owned_range;
+                break;
```
let's say we have:

non-owned ranges [1, 2], [4, 4]

owned ranges: [2, 2], [4, 5]


on first subtraction, we get range: [1, 1], and advance iterator to non-owned range [4, 4], whereas iterator to owned is still at [2, 2]

on next subtraction, we incorrectly subtract [2, 2] from [4, 4] instead of second owned range which is [4, 5].

in case 0, shouldn't we make that iterator to owned range is advanced until non-owned is before or contained in the current owned range?

the direction of implementation looks good to me (other than the possible problem above), but it was hard for me to grasp all details at this time of the night :-) I'll probably forget about it tomorrow. I think this is a good candidate for unit test, we could add a public static function in cleanup_compaction:

dht::partition_range_vector
    get_ranges_for_invalidation(partition_range_vector owned_ranges, partition_range_vector non_owned_ranges)

then use it in the unit test and cleanup itself.

another thing that came to my mind:
we can probably do some renaming to make the procedure easier to follow; I had a problem with the names being too similar. 
non_owned_ranges -> candidate_ranges
new_non_owned_ranges -> disowned_ranges or ranges_for_invalidation

bhalevy replied:
> let's say we have:
> 
> non-owned ranges [1, 2], [4, 4]
> 
> owned ranges: [2, 2], [4, 5]
> 
> on first subtraction, we get range: [1, 1], and advance iterator to non-owned range [4, 4], whereas iterator to owned is still at [2, 2]
> 
> on next subtraction, we incorrectly subtract [2, 2] from [4, 4] instead of second owned range which is [4, 5].

We shouldn't.  On next iteration, [2, 2] and [4, 4] do not intersect, hence [4, 4].subtract([2, 2]) = [4, 4],
so we fall into the size==1 case below, and since cmp(2, 4) < 0 we will set `*non_owned_range = std::move(ret[0]);` that will keep it the same, and increment `++owned_range`.

> 
> in case 0, shouldn't we make that iterator to owned range is advanced until non-owned is before or contained in the current owned range?
> 
> the direction of implementation looks good to me (other than the possible problem above), but it was hard for me to grasp all details at this time of the night :-) I'll probably forget about it tomorrow. I think this is a good candidate for unit test, we could add a public static function in cleanup_compaction:
> 
> dht::partition_range_vector get_ranges_for_invalidation(partition_range_vector owned_ranges, partition_range_vector non_owned_ranges)
> 
> then use it in the unit test and cleanup itself.

We can add a unit test that go over the results and make sure they are contained in full in the original non_owned_ranges and they don't intersect with owned_ranges.

> 
> another thing that came to my mind: we can probably do some renaming to make the procedure easier to follow; I had a problem with the names being too similar. non_owned_ranges -> candidate_ranges new_non_owned_ranges -> disowned_ranges or ranges_for_invalidation

no problem.

raphaelsc replied:
thanks for the explanation.

bhalevy commented:
In v5 ([f7a9dd7](https://github.com/scylladb/scylladb/commit/f7a9dd7c080d9f19e99ccd26a164c9770061e281)):
- compaction: refactor dht::subtract_ranges out of get_ranges_for_invalidation
- added unit test to verify implementations correctness

raphaelsc commented on compaction/compaction.cc:
```diff
@@ -1209,21 +1210,8 @@ class cleanup_compaction final : public regular_compaction {
-        }
-        return non_owned_ranges;
+
+        return dht::subtract_ranges(*_schema, non_owned_ranges, owned_ranges).get();
```
std::move(owned_ranges) ?

shouldn't block this series though.

raphaelsc commented on test/boost/sstable_compaction_test.cc:
```diff
@@ -5017,3 +5017,69 @@ SEASTAR_TEST_CASE(compaction_manager_stop_and_drain_race_test) {
+// dht::subtract_ranges is used by cleanup_compaction
+// get_ranges_for_invalidation.
+SEASTAR_THREAD_TEST_CASE(test_dht_subtract_ranges) {
+    simple_schema ss;
```
it can be moved to partitioned_test or range_test.

bhalevy replied:
I wanted to place it in range_test, but it is a a pure boost test and subtract_ranges is a coroutine.
partitioner_test looks good.

bhalevy commented on dht/i_partitioner.cc:
```diff
@@ -367,19 +367,69 @@ future<dht::partition_range_vector> subtract_ranges(const schema& schema, const
     auto cmp = dht::ring_position_comparator(schema);
```
yes

bhalevy commented:
In v6 ([6d54c32](https://github.com/scylladb/scylladb/commit/6d54c3241698121114042d13933bc3a0e531d9ae)):
- Added requirement for ranges_to_subtract to be sorted and deoverlapped to declaration comment
- std::move(owned_ranges) to dht::subtract_ranges
- moved unit test to partitioner_test

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/2951/

bhalevy commented:
> CI state **`FAILURE`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/2951/

faile on topology related assert failure in cdc test (cc @xemul)

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/2961/

bhalevy commented:
@scylladb/scylla-maint ping

denesb commented:
> @scylladb/scylla-maint ping

Conflicts, please rebase.

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/3093/
